polkadot-host-spec/ae-hostapi.tm

```diff
@@ -204,10 +204,38 @@
     root.
   </itemize>
 
-  <subsection|<verbatim|ext_storage_changes_root>>
+  <subsection|<verbatim|ext_storage_changes_root>><label|sect-ext-storage-changes-root>
 
-  Commits all existing operations and gets the resulting change root. The
-  parent hash is a SCALE encoded change root.
+  Commits all existing operations and gets the resulting change root as
+  described in section <reference|sect-changes-trie>. The parent hash is a
+  SCALE encoded change root.
+
+  \;
+
+  Unlike extrinsic mappings, block mappings are not created for each block.
+  The Runtime sets the key <verbatim|:changes_trie> in storage with its value
+  being a tuple of two 32-bit integers:
+
+  <\itemize-dot>
+    <item><verbatim|digest_interval> - The interval (in blocks) at which
+    block mappings are created. Block mappings are not created when this is
+    less or equal to <verbatim|1>.
+
+    <item><verbatim|digest_levels> - Maximal number of \Plevels\Q in the
```

Comment:
I don't really understand what hierarchy is meant here. Does that mean that if, for example, I have interval of 3 and levels of 4, I have to add to the changes trie of every 3rd block the entries about  changes made in past 3 blocks, to every 9th block changes made in past 9 blocks, to every 27th block for past 27 blocks and so on to 3^4? So I have to store the information about what keys were changed by which blocks for 81 blocks?

Reply:
Well, the nice thing about the way this spacing works is

Reply:
You only need to keep a handful of digest tries around. E.g. for `interval = 3` and `level = 4` the you would have a try over 3, 9, 27, 81 blocks. Whenever one of these intervals has passed, you add the current trie for this interval to the merkel prove. Then the trie of that interval is reset (I assume they are also added to a local database somewhere). If you want to save on computation you could update the higher intervals only when the interval just below it is reset. Because the intervals contain multiples of each other, you can use the already consolidated tries form the lower levels instead of adding each single transaction to each interval trie every single time.